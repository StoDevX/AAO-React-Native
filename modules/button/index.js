// @flow
import * as React from 'react'
import {StyleSheet, Platform} from 'react-native'
import BasicButton from 'react-native-button'
import noop from 'lodash/noop'
import {material, iOSUIKit} from 'react-native-typography'
import * as c from '@frogpond/colors'
import {type AppTheme} from '@frogpond/app-theme'
import {withTheme} from '@callstack/react-theme-provider'

const styles = StyleSheet.create({
	button: {
		alignSelf: 'center',
		paddingVertical: 10,
		paddingHorizontal: 20,
		marginVertical: 10,
		borderRadius: 6,
		overflow: 'hidden',
	},
	disabled: {
		backgroundColor: c.sectionBgColor,
	},
	text: {
		...Platform.select({
			ios: iOSUIKit.calloutWhiteObject,
			android: material.buttonWhiteObject,
		}),
	},
	textDisabled: {
		color: c.iosDisabledText,
	},
})

const inverted = StyleSheet.create({
	text: {
		...Platform.select({
			ios: iOSUIKit.calloutBlackObject,
			android: material.buttonBlackObject,
		}),
	},
})

type Props = {
	title?: string,
	onPress?: () => any,
	disabled?: boolean,
	buttonStyle?: any,
	textStyle?: any,
	mode?: 'default' | 'inverted',
	theme: AppTheme,
}

function Button({
	title = 'Push me!',
	onPress = noop,
	disabled = false,
	buttonStyle = null,
	textStyle = null,
	mode = 'default',
	theme,
}: Props) {
	let background =
		mode === 'default'
			? {backgroundColor: theme.buttonBackground}
			: {backgroundColor: theme.buttonForeground}
	let foreground =
		mode === 'default'
			? {color: theme.buttonForeground}
			: {color: theme.buttonBackground}

	let textStyleThing = mode === 'default' ? styles.text : inverted.text
	let containerStyle = [styles.button, background, buttonStyle]
	let style = [textStyleThing, foreground, textStyle]

	return (
		<BasicButton
			containerStyle={containerStyle}
			disabled={disabled}
			disabledContainerStyle={styles.disabled}
			onPress={onPress}
			style={style}
			styleDisabled={styles.textDisabled}
		>
			{Platform.OS === 'android' ? title.toUpperCase() : title}
		</BasicButton>
	)
}

export const RawButton = Button

const ThemedButton = withTheme(Button)

export {ThemedButton as Button}
