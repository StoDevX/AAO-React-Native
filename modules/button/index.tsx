import {
	Platform,
	StyleProp,
	StyleSheet,
	TextStyle,
	ViewStyle,
} from 'react-native'
import BasicButton from 'react-native-button'
import noop from 'lodash/noop'
import {iOSUIKit, material} from 'react-native-typography'
import * as c from '@frogpond/colors'
import {useTheme} from '@frogpond/app-theme'

const styles = StyleSheet.create({
	button: {
		alignSelf: 'center',
		paddingVertical: 10,
		paddingHorizontal: 20,
		marginVertical: 10,
		borderRadius: 6,
		overflow: 'hidden',
	},
	disabled: {
		backgroundColor: c.sectionBgColor,
	},
	text: {
		...Platform.select({
			ios: iOSUIKit.calloutWhiteObject,
			android: material.buttonWhiteObject,
		}),
	},
	textDisabled: {
		color: c.iosDisabledText,
	},
})

const inverted = StyleSheet.create({
	text: {
		...Platform.select({
			ios: iOSUIKit.calloutObject,
			android: material.buttonObject,
		}),
	},
})

type Props = {
	title?: string
	onPress?: () => unknown
	disabled?: boolean
	buttonStyle?: StyleProp<ViewStyle>
	textStyle?: StyleProp<TextStyle>
	mode?: 'default' | 'inverted'
}

export function Button({
	title = 'Push me!',
	onPress = noop,
	disabled = false,
	buttonStyle = null,
	textStyle = null,
	mode = 'default',
}: Props): JSX.Element {
	const theme = useTheme()

	const background =
		mode === 'default'
			? {backgroundColor: theme?.buttonBackground}
			: {backgroundColor: theme?.buttonForeground}
	const foreground =
		mode === 'default'
			? {color: theme?.buttonForeground}
			: {color: theme?.buttonBackground}

	const textStyleThing = mode === 'default' ? styles.text : inverted.text
	const containerStyle = [styles.button, background, buttonStyle]
	const style = [textStyleThing, foreground, textStyle]

	return (
		<BasicButton
			containerStyle={containerStyle}
			disabled={disabled}
			disabledContainerStyle={styles.disabled}
			onPress={onPress}
			style={style}
			styleDisabled={styles.textDisabled}
		>
			{Platform.OS === 'android' ? title.toUpperCase() : title}
		</BasicButton>
	)
}
