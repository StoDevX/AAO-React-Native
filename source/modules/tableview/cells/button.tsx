import * as React from 'react'
import {StyleProp, StyleSheet, Text, TextStyle} from 'react-native'
import {Cell} from '../index'
import * as c from '../../colors'
import Ionicon from '@expo/vector-icons/Ionicons'

type IconName = keyof typeof Ionicon.glyphMap

const styles = StyleSheet.create({
	title: {
		textAlign: 'left',
	},
	active: {
		color: c.link,
	},
	disabled: {
		color: c.secondaryLabel,
	},
})

interface ButtonCellProps {
	indeterminate?: boolean
	disabled?: boolean
	onPress: () => void
	textStyle?: StyleProp<TextStyle>
	title: string
	accessoryIcon?: IconName
}

export function ButtonCell({
	indeterminate = false,
	disabled = false,
	onPress,
	textStyle,
	title,
	accessoryIcon,
}: ButtonCellProps): React.JSX.Element {
	return (
		<Cell
			cellAccessoryView={
				accessoryIcon ? (
					<Ionicon
						color={disabled ? styles.disabled.color : styles.active.color}
						name={accessoryIcon}
						size={26}
					/>
				) : null
			}
			cellStyle="RightDetail"
			isDisabled={indeterminate || disabled}
			onPress={onPress}
			title={
				<Text
					style={[
						indeterminate || disabled ? styles.disabled : styles.active,
						textStyle,
					]}
				>
					{title}
				</Text>
			}
			titleTextStyle={styles.title}
		/>
	)
}
