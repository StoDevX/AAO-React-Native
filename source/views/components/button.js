// @flow
import * as React from 'react'
import {StyleSheet, Platform} from 'react-native'
import BasicButton from 'react-native-button'
import noop from 'lodash/noop'
import {material, iOSUIKit} from 'react-native-typography'

import * as c from './colors'

const styles = StyleSheet.create({
	button: {
		backgroundColor: c.denim,
		alignSelf: 'center',
		paddingVertical: 10,
		paddingHorizontal: 20,
		marginVertical: 10,
		borderRadius: 6,
		overflow: 'hidden',
	},
	disabled: {
		backgroundColor: c.iosLightBackground,
	},
	text: {
		...Platform.select({
			ios: iOSUIKit.calloutWhiteObject,
			android: material.buttonWhiteObject,
		}),
	},
	textDisabled: {
		color: c.iosDisabledText,
	},
})

type Props = {
	title?: string,
	onPress?: () => any,
	disabled?: boolean,
}

export function Button({
	title = 'Push me!',
	onPress = noop,
	disabled = false,
}: Props) {
	return (
		<BasicButton
			containerStyle={styles.button}
			disabled={disabled}
			disabledContainerStyle={styles.disabled}
			onPress={onPress}
			style={styles.text}
			styleDisabled={styles.textDisabled}
		>
			{Platform.OS === 'android' ? title.toUpperCase() : title}
		</BasicButton>
	)
}
