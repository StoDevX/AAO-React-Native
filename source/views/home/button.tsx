import * as React from 'react'
import {Platform, StyleSheet, Text, View} from 'react-native'
import Entypo from '@expo/vector-icons/Entypo'
import {Touchable} from '../../../modules/touchable'
import {transparent} from '../../../modules/colors'
import {homescreenForegroundDark, homescreenForegroundLight} from './colors'

type Props = {
	title: string
	iconName: keyof (typeof Entypo)['glyphMap']
	foreground: 'light' | 'dark'
	tintColor: string
}

export function HomeScreenButton(props: Props): JSX.Element {
	let {title, iconName, foreground, tintColor} = props
	let foregroundStyle =
		foreground === 'light' ? styles.lightForeground : styles.darkForeground

	return (
		<Touchable
			accessibilityLabel={title}
			accessibilityRole="button"
			accessible={true}
			highlight={false}
			style={[styles.button, {backgroundColor: tintColor}]}
		>
			<View style={styles.contents}>
				<Entypo
					name={iconName}
					size={32}
					style={[foregroundStyle, styles.icon]}
				/>
				<Text style={[foregroundStyle, styles.text]}>{title}</Text>
			</View>
		</Touchable>
	)
}

export const CELL_MARGIN = 10
const cellVerticalPadding = 8
const cellHorizontalPadding = 4
// hasNotch() ? 17 : 6

const styles = StyleSheet.create({
	button: {
		elevation: 2,
		borderRadius: Platform.OS === 'ios' ? 17 : 3,

		marginBottom: CELL_MARGIN,
		marginLeft: CELL_MARGIN / 2,
		marginRight: CELL_MARGIN / 2,
	},
	contents: {
		alignItems: 'center',
		justifyContent: 'center',

		paddingTop: cellVerticalPadding,
		paddingBottom: cellVerticalPadding / 2,
		paddingHorizontal: cellHorizontalPadding,
	},
	icon: {
		backgroundColor: transparent,
	},
	text: {
		backgroundColor: transparent,
		fontFamily: Platform.OS === 'ios' ? 'System' : 'sans-serif-condensed',
		fontSize: 14,
	},
	lightForeground: {
		color: homescreenForegroundLight,
	},
	darkForeground: {
		color: homescreenForegroundDark,
	},
})
