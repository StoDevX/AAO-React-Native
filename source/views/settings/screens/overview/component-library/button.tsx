import * as React from 'react'
import {Alert} from 'react-native'

import * as c from '@frogpond/colors'
import {Section, Cell} from '@frogpond/tableview'
import {Button} from '@frogpond/button'
import {ButtonCell} from '@frogpond/tableview/cells'
import {LibraryWrapper} from './base/library-wrapper'

interface RowProps {
	title: string
	children: JSX.Element
}

const Example = ({title, children}: RowProps): JSX.Element => {
	return (
		<Cell cellAccessoryView={children} cellStyle="RightDetail" title={title} />
	)
}

const ButtonCellExample = (): JSX.Element => {
	return (
		<>
			<ButtonCell disabled={false} onPress={() => undefined} title="Enabled" />

			<ButtonCell disabled={true} onPress={() => undefined} title="Disabled" />

			<ButtonCell
				indeterminate={true}
				onPress={() => undefined}
				title="Indeterminate"
			/>

			<ButtonCell
				accessoryIcon="school"
				onPress={() => undefined}
				title="Accessory"
			/>

			<ButtonCell
				accessoryIcon="school"
				onPress={() => undefined}
				textStyle={{color: c.red}}
				title="Accessory, textstyle"
			/>

			<ButtonCell
				accessoryIcon="school"
				disabled={true}
				onPress={() => undefined}
				textStyle={{color: c.red}}
				title="Disabled, textstyle, accessory"
			/>

			<ButtonCell
				disabled={false}
				onPress={() => Alert.alert('You tapped the button!')}
				title="Callback"
			/>
		</>
	)
}

const ButtonExample = (): JSX.Element => {
	return (
		<>
			<Example title="No props">
				<Button />
			</Example>

			<Example title="Enabled">
				<Button title="Tap me" />
			</Example>

			<Example title="Disabled">
				<Button disabled={true} title="Tap me" />
			</Example>

			<Example title="Callback">
				<Button
					onPress={() => Alert.alert('You tapped the button!')}
					title="Tap me"
				/>
			</Example>

			<Example title="Inverted (fix me)">
				<Button mode="inverted" title="Tap me" />
			</Example>

			<Example title="Truncated (fix me)">
				<Button title="Very long button text that should truncate and not wrap to the next line" />
			</Example>
		</>
	)
}

export const ButtonLibrary = (): JSX.Element => (
	<LibraryWrapper>
		<>
			<Section header="@frogpond/tableview/cells">
				<ButtonCellExample />
			</Section>

			<Section header="@frogpond/button">
				<ButtonExample />
			</Section>
		</>
	</LibraryWrapper>
)
