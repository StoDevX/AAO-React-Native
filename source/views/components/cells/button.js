// @flow

import * as React from 'react'
import {StyleSheet, Text} from 'react-native'
import {Cell} from 'react-native-tableview-simple'
import * as c from '../../components/colors'

const styles = StyleSheet.create({
	title: {
		textAlign: 'left',
	},
	active: {
		color: c.infoBlue,
	},
	disabled: {
		color: c.iosDisabledText,
	},
})

export function ButtonCell({
	indeterminate,
	disabled,
	onPress,
	title,
}: {
	indeterminate?: boolean,
	disabled?: boolean,
	onPress: () => any,
	title: string,
}) {
	return (
		<Cell
			isDisabled={indeterminate || disabled}
			onPress={onPress}
			title={
				<Text
					style={[indeterminate || disabled ? styles.disabled : styles.active]}
				>
					{title}
				</Text>
			}
			titleTextStyle={styles.title}
		/>
	)
}
