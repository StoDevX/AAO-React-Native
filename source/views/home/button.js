// @flow

import React from 'react'
import {Text, StyleSheet, Platform, Dimensions} from 'react-native'
import Icon from 'react-native-vector-icons/Entypo'
import type {ViewType} from '../views'
import {Touchable} from '../components/touchable'
import * as c from '../components/colors'

const Viewport = Dimensions.get('window')

export function HomeScreenButton({
  view,
  onPress,
}: {
  view: ViewType,
  onPress: () => any,
}) {
  return (
    <Touchable
      highlight={false}
      onPress={onPress}
      style={[styles.rectangle, {backgroundColor: view.tint}]}
    >
      <Icon name={view.icon} size={32} style={styles.rectangleButtonIcon} />
      <Text style={styles.rectangleButtonText}>
        {view.title}
      </Text>
    </Touchable>
  )
}

export const CELL_MARGIN = 10
const cellVerticalPadding = 8
const cellHorizontalPadding = 4
const cellWidth = Viewport.width / 2 - CELL_MARGIN * 1.5

const styles = StyleSheet.create({
  // Main buttons for actions on home screen
  rectangle: {
    width: cellWidth,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: cellVerticalPadding,
    paddingBottom: cellVerticalPadding / 2,
    paddingHorizontal: cellHorizontalPadding,
    borderRadius: Platform.OS === 'ios' ? 6 : 3,
    elevation: 2,

    marginTop: CELL_MARGIN / 2,
    marginBottom: CELL_MARGIN / 2,
    marginLeft: CELL_MARGIN / 2,
    marginRight: CELL_MARGIN / 2,
  },

  // Text styling in buttons
  rectangleButtonIcon: {
    color: c.white,
  },
  rectangleButtonText: {
    color: c.white,
    marginTop: cellVerticalPadding / 2,
    fontFamily: Platform.OS === 'ios' ? 'System' : 'sans-serif-condensed',
    textAlign: 'center',
    fontSize: 14,
  },
})
