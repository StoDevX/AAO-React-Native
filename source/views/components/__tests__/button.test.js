/* eslint-env jest */
// @flow

import 'react-native'
import * as React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'

import {Button} from '../button'

const shallow = component => {
  const r = new ReactShallowRenderer()
  r.render(component)
  return r.getRenderOutput()
}

test('renders', () => {
  const tree = shallow(<Button />)
  expect(tree).toMatchSnapshot()
})

test('can change the title', () => {
  const tree = shallow(<Button title="Title" />)
  expect(tree).toMatchSnapshot()
})

xtest('calls the callback', () => {
  const cb = jest.fn()

  const tree = shallow(<Button onPress={cb} />)
  expect(tree).toMatchSnapshot()

  tree.simulate('press')

  expect(cb).toHaveBeenCalled()
  expect(tree).toMatchSnapshot()
})
