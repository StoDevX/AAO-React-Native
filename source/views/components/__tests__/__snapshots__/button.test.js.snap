// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls the callback 1`] = `
<Button
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "backgroundColor": "rgb(67, 114, 170)",
        "borderRadius": 6,
        "color": "#FFFFFF",
        "marginVertical": 10,
        "overflow": "hidden",
        "paddingVertical": 10,
        "width": 200,
      },
      false,
    ]
  }
>
  Push me!
</Button>
`;

exports[`calls the callback 2`] = `
<Button
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "backgroundColor": "rgb(67, 114, 170)",
        "borderRadius": 6,
        "color": "#FFFFFF",
        "marginVertical": 10,
        "overflow": "hidden",
        "paddingVertical": 10,
        "width": 200,
      },
      false,
    ]
  }
>
  Push me!
</Button>
`;

exports[`can change the title 1`] = `
<Button
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "backgroundColor": "rgb(67, 114, 170)",
        "borderRadius": 6,
        "color": "#FFFFFF",
        "marginVertical": 10,
        "overflow": "hidden",
        "paddingVertical": 10,
        "width": 200,
      },
      false,
    ]
  }
>
  Title
</Button>
`;

exports[`renders 1`] = `
<Button
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "backgroundColor": "rgb(67, 114, 170)",
        "borderRadius": 6,
        "color": "#FFFFFF",
        "marginVertical": 10,
        "overflow": "hidden",
        "paddingVertical": 10,
        "width": 200,
      },
      false,
    ]
  }
>
  Push me!
</Button>
`;
